File: by-language/python/data_structures/stack.py
"""
Stack (LIFO - Last In, First Out)

╔════════════════════════════════════════════════════════════════════════════╗
║                           VISUAL REPRESENTATION                            ║
╠════════════════════════════════════════════════════════════════════════════╣
║                                                                            ║
║  Stack Structure:                                                          ║
║  ┌─────────────────┐                                                       ║
║  │       TOP       │  ◄── push() and pop() happen here                     ║
║  ├─────────────────┤                                                       ║
║  │       40        │  ◄── most recently added (top)                        ║
║  ├─────────────────┤                                                       ║
║  │       30        │                                                       ║
║  ├─────────────────┤                                                       ║
║  │       20        │                                                       ║
║  ├─────────────────┤                                                       ║
║  │       10        │  ◄── first added (bottom)                             ║
║  └─────────────────┘                                                       ║
║      BOTTOM                                                                ║
║                                                                            ║
║  Push Operation:                                                           ║
║  Before: [10, 20, 30]  (top = 30)                                          ║
║  push(40)                                                                  ║
║  After:  [10, 20, 30, 40]  (top = 40)                                      ║
║                                                                            ║
║  Pop Operation:                                                            ║
║  Before: [10, 20, 30, 40]  (top = 40)                                      ║
║  pop() → returns 40                                                        ║
║  After:  [10, 20, 30]  (top = 30)                                          ║
║                                                                            ║
║  Min Stack Visualization:                                                  ║
║  ┌─────────────────┬─────────────────┐                                     ║
║  │  Main Stack     │   Min Stack     │                                     ║
║  ├─────────────────┼─────────────────┤                                     ║
║  │      5          │      5          │  ◄── push 5: min = 5                ║
║  │      3          │      3          │  ◄── push 3: min = 3                ║
║  │      7          │      3          │  ◄── push 7: min still 3            ║
║  │      2          │      2          │  ◄── push 2: min = 2                ║
║  └─────────────────┴─────────────────┘                                     ║
║                                                                            ║
╚════════════════════════════════════════════════════════════════════════════╝

COMPLEXITY:
┌─────────────────────┬─────────────┬─────────────┬─────────────┐
│ Operation           │ Time        │ Space       │ Notes       │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ push                │ O(1)*       │ O(1)        │ *amortized  │
│ pop                 │ O(1)        │ O(1)        │             │
│ peek/top            │ O(1)        │ O(1)        │             │
│ is_empty            │ O(1)        │ O(1)        │             │
│ size                │ O(1)        │ O(1)        │             │
│ get_min (MinStack)  │ O(1)        │ O(n)        │ extra stack │
│ get_max (MaxStack)  │ O(1)        │ O(n)        │ extra stack │
└─────────────────────┴─────────────┴─────────────┴─────────────┘

LEETCODE PROBLEMS:
- #20 Valid Parentheses
- #155 Min Stack
- #232 Implement Queue using Stacks
- #225 Implement Stack using Queues
- #150 Evaluate Reverse Polish Notation
- #71 Simplify Path
- #84 Largest Rectangle in Histogram
- #85 Maximal Rectangle
- #739 Daily Temperatures
- #496 Next Greater Element I
- #503 Next Greater Element II
- #42 Trapping Rain Water

USE CASES:
- Function call stack (recursion)
- Undo/Redo operations
- Browser back/forward buttons
- Expression evaluation (postfix, prefix)
- Parentheses matching
- Depth-first search (DFS)
- Backtracking algorithms
- Syntax parsing
"""

from typing import TypeVar, Generic, Iterator, Optional, List

T = TypeVar('T')


class Stack(Generic[T]):
    """
    A stack implementation using a dynamic array (list).

    This is the standard stack with O(1) push and pop operations.
    Uses Python's list which provides amortized O(1) append.
    """

    def __init__(self) -> None:
        """Initialize an empty stack."""
        self._data: List[T] = []

    def __len__(self) -> int:
        """Return the number of elements in the stack."""
        return len(self._data)

    def __iter__(self) -> Iterator[T]:
        """Iterate over elements from bottom to top."""
        return iter(self._data)

    def __repr__(self) -> str:
        """Return string representation of the stack."""
        return f"Stack({self._data})"

    def __eq__(self, other: object) -> bool:
        """Check equality with another stack."""
        if not isinstance(other, Stack):
            return NotImplemented
        return self._data == other._data

    def __contains__(self, value: T) -> bool:
        """Check if value exists in stack."""
        return value in self._data

    @property
    def is_empty(self) -> bool:
        """Return True if stack is empty."""
        return len(self._data) == 0

    def push(self, value: T) -> None:
        """
        Push an element onto the stack.

        Time Complexity: O(1) amortized

        Args:
            value: The value to push.
        """
        self._data.append(value)

    def pop(self) -> T:
        """
        Pop and return the top element.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Pop from empty stack")
        return self._data.pop()

    def peek(self) -> T:
        """
        Return the top element without removing it.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty stack")
        return self._data[-1]

    def clear(self) -> None:
        """Remove all elements from the stack."""
        self._data.clear()

    def to_list(self) -> List[T]:
        """Convert stack to list (bottom to top)."""
        return self._data.copy()

    @classmethod
    def from_list(cls, items: List[T]) -> "Stack[T]":
        """
        Create a stack from a list.

        First element of list becomes bottom of stack.

        Args:
            items: List of items.

        Returns:
            A new Stack.
        """
        stack: Stack[T] = cls()
        for item in items:
            stack.push(item)
        return stack


class MinStack(Generic[T]):
    """
    A stack that supports O(1) retrieval of minimum element.

    Uses an auxiliary stack to track minimums.
    """

    def __init__(self) -> None:
        """Initialize an empty min stack."""
        self._data: List[T] = []
        self._min_stack: List[T] = []

    def __len__(self) -> int:
        """Return the number of elements in the stack."""
        return len(self._data)

    def __repr__(self) -> str:
        """Return string representation."""
        return f"MinStack({self._data})"

    @property
    def is_empty(self) -> bool:
        """Return True if stack is empty."""
        return len(self._data) == 0

    def push(self, value: T) -> None:
        """
        Push an element onto the stack.

        Time Complexity: O(1)

        Args:
            value: The value to push.
        """
        self._data.append(value)

        if not self._min_stack or value <= self._min_stack[-1]:
            self._min_stack.append(value)

    def pop(self) -> T:
        """
        Pop and return the top element.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Pop from empty stack")

        value = self._data.pop()

        if value == self._min_stack[-1]:
            self._min_stack.pop()

        return value

    def peek(self) -> T:
        """
        Return the top element without removing it.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty stack")
        return self._data[-1]

    def get_min(self) -> T:
        """
        Return the minimum element in the stack.

        Time Complexity: O(1)

        Returns:
            The minimum element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("get_min from empty stack")
        return self._min_stack[-1]


class MaxStack(Generic[T]):
    """
    A stack that supports O(1) retrieval of maximum element.

    Uses an auxiliary stack to track maximums.
    """

    def __init__(self) -> None:
        """Initialize an empty max stack."""
        self._data: List[T] = []
        self._max_stack: List[T] = []

    def __len__(self) -> int:
        """Return the number of elements in the stack."""
        return len(self._data)

    def __repr__(self) -> str:
        """Return string representation."""
        return f"MaxStack({self._data})"

    @property
    def is_empty(self) -> bool:
        """Return True if stack is empty."""
        return len(self._data) == 0

    def push(self, value: T) -> None:
        """
        Push an element onto the stack.

        Time Complexity: O(1)

        Args:
            value: The value to push.
        """
        self._data.append(value)

        if not self._max_stack or value >= self._max_stack[-1]:
            self._max_stack.append(value)

    def pop(self) -> T:
        """
        Pop and return the top element.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Pop from empty stack")

        value = self._data.pop()

        if value == self._max_stack[-1]:
            self._max_stack.pop()

        return value

    def peek(self) -> T:
        """
        Return the top element without removing it.

        Time Complexity: O(1)

        Returns:
            The top element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty stack")
        return self._data[-1]

    def get_max(self) -> T:
        """
        Return the maximum element in the stack.

        Time Complexity: O(1)

        Returns:
            The maximum element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty:
            raise IndexError("get_max from empty stack")
        return self._max_stack[-1]


# Utility functions that use stacks


def is_balanced_parentheses(s: str) -> bool:
    """
    Check if parentheses in a string are balanced.

    Supports: (), [], {}

    Time Complexity: O(n)
    Space Complexity: O(n)

    Args:
        s: String containing parentheses.

    Returns:
        True if balanced, False otherwise.
    """
    stack: Stack[str] = Stack()
    pairs = {')': '(', ']': '[', '}': '{'}

    for char in s:
        if char in '([{':
            stack.push(char)
        elif char in ')]}':
            if stack.is_empty or stack.pop() != pairs[char]:
                return False

    return stack.is_empty


def evaluate_postfix(expression: str) -> float:
    """
    Evaluate a postfix (Reverse Polish Notation) expression.

    Time Complexity: O(n)
    Space Complexity: O(n)

    Args:
        expression: Space-separated postfix expression.
                   Example: "2 3 + 4 *" = (2 + 3) * 4 = 20

    Returns:
        The result of the expression.

    Raises:
        ValueError: If the expression is invalid.
    """
    stack: Stack[float] = Stack()
    operators = {'+', '-', '*', '/'}

    tokens = expression.split()

    for token in tokens:
        if token in operators:
            if len(stack) < 2:
                raise ValueError("Invalid expression")

            b = stack.pop()
            a = stack.pop()

            if token == '+':
                stack.push(a + b)
            elif token == '-':
                stack.push(a - b)
            elif token == '*':
                stack.push(a * b)
            elif token == '/':
                if b == 0:
                    raise ValueError("Division by zero")
                stack.push(a / b)
        else:
            try:
                stack.push(float(token))
            except ValueError:
                raise ValueError(f"Invalid token: {token}")

    if len(stack) != 1:
        raise ValueError("Invalid expression")

    return stack.pop()


def next_greater_element(nums: List[int]) -> List[int]:
    """
    Find the next greater element for each element in the array.

    For each element, find the first greater element to its right.
    If none exists, use -1.

    Time Complexity: O(n)
    Space Complexity: O(n)

    Args:
        nums: List of integers.

    Returns:
        List where result[i] is the next greater element for nums[i].

    Example:
        [4, 5, 2, 10, 8] -> [5, 10, 10, -1, -1]
    """
    n = len(nums)
    result = [-1] * n
    stack: Stack[int] = Stack()  # Stack of indices

    for i in range(n):
        while not stack.is_empty and nums[stack.peek()] < nums[i]:
            idx = stack.pop()
            result[idx] = nums[i]
        stack.push(i)

    return result

File: by-language/rust/dsa-data-structures/src/linear/stack.rs
//! Stack (LIFO - Last In, First Out)
//!
//! A stack is a linear data structure that follows the LIFO principle.
//!
//! ```text
//! ╔════════════════════════════════════════════════════════════════════════════╗
//! ║                           VISUAL REPRESENTATION                            ║
//! ╠════════════════════════════════════════════════════════════════════════════╣
//! ║                                                                            ║
//! ║  Stack Structure:                                                          ║
//! ║  ┌─────────┐                                                               ║
//! ║  │    5    │ ◀── top (push/pop here)                                       ║
//! ║  ├─────────┤                                                               ║
//! ║  │    4    │                                                               ║
//! ║  ├─────────┤                                                               ║
//! ║  │    3    │                                                               ║
//! ║  ├─────────┤                                                               ║
//! ║  │    2    │                                                               ║
//! ║  ├─────────┤                                                               ║
//! ║  │    1    │ ◀── bottom                                                    ║
//! ║  └─────────┘                                                               ║
//! ║                                                                            ║
//! ║  Push Operation:              Pop Operation:                               ║
//! ║  ┌─────────┐                  ┌─────────┐                                  ║
//! ║  │   NEW   │ ◀── push(6)      │    X    │ ──▶ pop() returns 5              ║
//! ║  ├─────────┤                  ├─────────┤                                  ║
//! ║  │    5    │                  │    4    │ ◀── new top                      ║
//! ║  ├─────────┤                  ├─────────┤                                  ║
//! ║  │   ...   │                  │   ...   │                                  ║
//! ║  └─────────┘                  └─────────┘                                  ║
//! ║                                                                            ║
//! ╚════════════════════════════════════════════════════════════════════════════╝
//! ```
//!
//! ## Complexity
//!
//! | Operation | Average | Worst | Space |
//! |-----------|---------|-------|-------|
//! | Push      | O(1)*   | O(n)  | O(1)  |
//! | Pop       | O(1)    | O(1)  | O(1)  |
//! | Peek      | O(1)    | O(1)  | O(1)  |
//! | Search    | O(n)    | O(n)  | O(1)  |
//! | Is Empty  | O(1)    | O(1)  | O(1)  |
//!
//! *Amortized O(1) due to occasional resizing
//!
//! ## LeetCode Problems
//!
//! - [#20 Valid Parentheses](https://leetcode.com/problems/valid-parentheses/)
//! - [#155 Min Stack](https://leetcode.com/problems/min-stack/)
//! - [#232 Implement Queue using Stacks](https://leetcode.com/problems/implement-queue-using-stacks/)
//! - [#150 Evaluate Reverse Polish Notation](https://leetcode.com/problems/evaluate-reverse-polish-notation/)
//! - [#739 Daily Temperatures](https://leetcode.com/problems/daily-temperatures/)
//! - [#84 Largest Rectangle in Histogram](https://leetcode.com/problems/largest-rectangle-in-histogram/)
//! - [#71 Simplify Path](https://leetcode.com/problems/simplify-path/)
//! - [#394 Decode String](https://leetcode.com/problems/decode-string/)
//!
//! ## Use Cases
//!
//! - Function call stack (recursion)
//! - Expression evaluation and parsing
//! - Undo/redo operations
//! - Backtracking algorithms
//! - Browser history (back button)
//! - Balanced parentheses checking
//!
//! ## Example
//!
//! ```rust
//! use dsa_data_structures::linear::Stack;
//!
//! let mut stack = Stack::new();
//! stack.push(1);
//! stack.push(2);
//! stack.push(3);
//!
//! assert_eq!(stack.len(), 3);
//! assert_eq!(stack.peek(), Some(&3));
//! assert_eq!(stack.pop(), Some(3));
//! assert_eq!(stack.pop(), Some(2));
//! ```

use alloc::vec::Vec;

use dsa_core::{Container, DsaError, Result, Searchable};

/// A LIFO (Last In, First Out) stack implementation.
///
/// This implementation uses a `Vec` as the underlying storage,
/// providing amortized O(1) push and O(1) pop operations.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Stack<T> {
    data: Vec<T>,
}

impl<T> Stack<T> {
    /// Creates a new empty stack.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Stack;
    ///
    /// let stack: Stack<i32> = Stack::new();
    /// assert!(stack.is_empty());
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Stack { data: Vec::new() }
    }

    /// Creates a new stack with the specified capacity.
    ///
    /// # Time Complexity
    /// O(1)
    #[must_use]
    pub fn with_capacity(capacity: usize) -> Self {
        Stack {
            data: Vec::with_capacity(capacity),
        }
    }

    /// Returns the number of elements in the stack.
    ///
    /// # Time Complexity
    /// O(1)
    #[inline]
    #[must_use]
    pub fn len(&self) -> usize {
        self.data.len()
    }

    /// Returns `true` if the stack contains no elements.
    ///
    /// # Time Complexity
    /// O(1)
    #[inline]
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    /// Pushes an element onto the top of the stack.
    ///
    /// # Time Complexity
    /// O(1) amortized, O(n) worst case when resizing.
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Stack;
    ///
    /// let mut stack = Stack::new();
    /// stack.push(10);
    /// stack.push(20);
    /// assert_eq!(stack.peek(), Some(&20));
    /// ```
    pub fn push(&mut self, value: T) {
        self.data.push(value);
    }

    /// Removes and returns the top element from the stack.
    ///
    /// Returns `None` if the stack is empty.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Stack;
    ///
    /// let mut stack = Stack::new();
    /// stack.push(10);
    /// stack.push(20);
    /// assert_eq!(stack.pop(), Some(20));
    /// assert_eq!(stack.pop(), Some(10));
    /// assert_eq!(stack.pop(), None);
    /// ```
    pub fn pop(&mut self) -> Option<T> {
        self.data.pop()
    }

    /// Returns a reference to the top element without removing it.
    ///
    /// Returns `None` if the stack is empty.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Stack;
    ///
    /// let mut stack = Stack::new();
    /// stack.push(10);
    /// assert_eq!(stack.peek(), Some(&10));
    /// assert_eq!(stack.len(), 1); // Element not removed
    /// ```
    #[must_use]
    pub fn peek(&self) -> Option<&T> {
        self.data.last()
    }

    /// Returns a mutable reference to the top element without removing it.
    ///
    /// # Time Complexity
    /// O(1)
    #[must_use]
    pub fn peek_mut(&mut self) -> Option<&mut T> {
        self.data.last_mut()
    }

    /// Clears the stack, removing all elements.
    ///
    /// # Time Complexity
    /// O(n)
    pub fn clear(&mut self) {
        self.data.clear();
    }

    /// Returns an iterator over the stack from bottom to top.
    pub fn iter(&self) -> impl Iterator<Item = &T> {
        self.data.iter()
    }

    /// Returns an iterator over the stack from top to bottom.
    pub fn iter_from_top(&self) -> impl Iterator<Item = &T> {
        self.data.iter().rev()
    }

    /// Converts the stack to a `Vec` (bottom to top).
    #[must_use]
    pub fn to_vec(&self) -> Vec<T>
    where
        T: Clone,
    {
        self.data.clone()
    }

    /// Creates a stack from a `Vec` (first element is bottom).
    #[must_use]
    pub fn from_vec(vec: Vec<T>) -> Self {
        Stack { data: vec }
    }
}

impl<T: PartialEq> Stack<T> {
    /// Searches for an element and returns its position from the top (1-indexed).
    ///
    /// Returns `None` if the element is not found.
    ///
    /// # Time Complexity
    /// O(n)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Stack;
    ///
    /// let stack = Stack::from_vec(vec![1, 2, 3, 4, 5]);
    /// assert_eq!(stack.search(&5), Some(1)); // Top element
    /// assert_eq!(stack.search(&1), Some(5)); // Bottom element
    /// assert_eq!(stack.search(&10), None);   // Not found
    /// ```
    #[must_use]
    pub fn search(&self, value: &T) -> Option<usize> {
        self.data
            .iter()
            .rev()
            .position(|x| x == value)
            .map(|pos| pos + 1)
    }

    /// Returns `true` if the stack contains the specified value.
    ///
    /// # Time Complexity
    /// O(n)
    #[must_use]
    pub fn contains(&self, value: &T) -> bool {
        self.data.contains(value)
    }
}

impl<T> Container for Stack<T> {
    fn len(&self) -> usize {
        self.data.len()
    }
}

impl<T: PartialEq> Searchable<T> for Stack<T> {
    fn search(&self, value: &T) -> Option<usize> {
        // Returns 0-indexed position from bottom for trait compatibility
        self.data.iter().position(|x| x == value)
    }
}

impl<T> Default for Stack<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> FromIterator<T> for Stack<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        Stack {
            data: iter.into_iter().collect(),
        }
    }
}

impl<T> IntoIterator for Stack<T> {
    type Item = T;
    type IntoIter = alloc::vec::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.data.into_iter()
    }
}

impl<'a, T> IntoIterator for &'a Stack<T> {
    type Item = &'a T;
    type IntoIter = core::slice::Iter<'a, T>;

    fn into_iter(self) -> Self::IntoIter {
        self.data.iter()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod basics {
        use super::*;

        #[test]
        fn test_new() {
            let stack: Stack<i32> = Stack::new();
            assert!(stack.is_empty());
            assert_eq!(stack.len(), 0);
        }

        #[test]
        fn test_with_capacity() {
            let stack: Stack<i32> = Stack::with_capacity(10);
            assert!(stack.is_empty());
        }

        #[test]
        fn test_default() {
            let stack: Stack<i32> = Stack::default();
            assert!(stack.is_empty());
        }

        #[test]
        fn test_from_vec() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            assert_eq!(stack.len(), 3);
            assert_eq!(stack.peek(), Some(&3));
        }

        #[test]
        fn test_to_vec() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            assert_eq!(stack.to_vec(), vec![1, 2, 3]);
        }

        #[test]
        fn test_clone() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            let cloned = stack.clone();
            assert_eq!(stack, cloned);
        }
    }

    mod push_pop {
        use super::*;

        #[test]
        fn test_push() {
            let mut stack = Stack::new();
            stack.push(1);
            stack.push(2);
            stack.push(3);
            assert_eq!(stack.len(), 3);
            assert_eq!(stack.peek(), Some(&3));
        }

        #[test]
        fn test_pop() {
            let mut stack = Stack::from_vec(vec![1, 2, 3]);
            assert_eq!(stack.pop(), Some(3));
            assert_eq!(stack.pop(), Some(2));
            assert_eq!(stack.pop(), Some(1));
            assert_eq!(stack.pop(), None);
        }

        #[test]
        fn test_lifo_order() {
            let mut stack = Stack::new();
            stack.push(1);
            stack.push(2);
            stack.push(3);

            // Should come out in reverse order (LIFO)
            assert_eq!(stack.pop(), Some(3));
            assert_eq!(stack.pop(), Some(2));
            assert_eq!(stack.pop(), Some(1));
        }
    }

    mod peek {
        use super::*;

        #[test]
        fn test_peek() {
            let mut stack = Stack::new();
            assert_eq!(stack.peek(), None);

            stack.push(10);
            assert_eq!(stack.peek(), Some(&10));
            assert_eq!(stack.len(), 1); // Not removed

            stack.push(20);
            assert_eq!(stack.peek(), Some(&20));
        }

        #[test]
        fn test_peek_mut() {
            let mut stack = Stack::from_vec(vec![1, 2, 3]);
            if let Some(top) = stack.peek_mut() {
                *top = 30;
            }
            assert_eq!(stack.peek(), Some(&30));
        }
    }

    mod search {
        use super::*;

        #[test]
        fn test_search() {
            let stack = Stack::from_vec(vec![1, 2, 3, 4, 5]);
            assert_eq!(stack.search(&5), Some(1)); // Top
            assert_eq!(stack.search(&4), Some(2));
            assert_eq!(stack.search(&1), Some(5)); // Bottom
            assert_eq!(stack.search(&10), None);
        }

        #[test]
        fn test_contains() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            assert!(stack.contains(&2));
            assert!(!stack.contains(&5));
        }
    }

    mod utilities {
        use super::*;

        #[test]
        fn test_clear() {
            let mut stack = Stack::from_vec(vec![1, 2, 3]);
            stack.clear();
            assert!(stack.is_empty());
        }

        #[test]
        fn test_iter() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            let collected: Vec<_> = stack.iter().cloned().collect();
            assert_eq!(collected, vec![1, 2, 3]); // Bottom to top
        }

        #[test]
        fn test_iter_from_top() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            let collected: Vec<_> = stack.iter_from_top().cloned().collect();
            assert_eq!(collected, vec![3, 2, 1]); // Top to bottom
        }

        #[test]
        fn test_from_iter() {
            let stack: Stack<i32> = (1..=5).collect();
            assert_eq!(stack.len(), 5);
            assert_eq!(stack.peek(), Some(&5));
        }

        #[test]
        fn test_into_iter() {
            let stack = Stack::from_vec(vec![1, 2, 3]);
            let collected: Vec<_> = stack.into_iter().collect();
            assert_eq!(collected, vec![1, 2, 3]);
        }
    }

    mod edge_cases {
        use super::*;

        #[test]
        fn test_empty_operations() {
            let mut stack: Stack<i32> = Stack::new();
            assert_eq!(stack.pop(), None);
            assert_eq!(stack.peek(), None);
            assert!(!stack.contains(&1));
        }

        #[test]
        fn test_single_element() {
            let mut stack = Stack::new();
            stack.push(42);
            assert_eq!(stack.peek(), Some(&42));
            assert_eq!(stack.pop(), Some(42));
            assert!(stack.is_empty());
        }

        #[test]
        fn test_large_stack() {
            let mut stack = Stack::new();
            for i in 0..10000 {
                stack.push(i);
            }
            assert_eq!(stack.len(), 10000);
            assert_eq!(stack.peek(), Some(&9999));
        }
    }

    mod equality {
        use super::*;

        #[test]
        fn test_eq() {
            let s1 = Stack::from_vec(vec![1, 2, 3]);
            let s2 = Stack::from_vec(vec![1, 2, 3]);
            let s3 = Stack::from_vec(vec![1, 2, 4]);
            assert_eq!(s1, s2);
            assert_ne!(s1, s3);
        }
    }
}
